web_src/js/utils/dom.js

```diff
@@ -0,0 +1,65 @@
+function getComputedStyleProperty(el, prop) {
+  const cs = el ? window.getComputedStyle(el) : null;
+  return cs ? cs[prop] : null;
+}
+
+function isShown(el) {
+  return getComputedStyleProperty(el, 'display') !== 'none';
+}
+
+function assertShown(el, expectShown) {
+  if (window.config.runModeIsProd) return;
+
+  // to help developers to catch display bugs, this assertion can be removed after next release cycle or if it has been proved that there is no bug.
+  if (expectShown && !isShown(el)) {
+    throw new Error('element is hidden but should be shown');
+  } else if (!expectShown && isShown(el)) {
+    throw new Error('element is shown but should be hidden');
+  }
+}
+
+function elementsCall(el, func, ...args) {
+  if (el instanceof String) {
+    el = document.querySelectorAll(el);
+  }
+  if (el instanceof Node) {
+    func(el, ...args);
+  } else if (el.length !== undefined) {
+    // this works for: NodeList, HTMLCollection, Array, jQuery
+    for (const e of el) {
+      func(e, ...args);
+    }
+  } else {
+    throw new Error('invalid argument to be shown/hidden');
+  }
+}
+
+function toggleShown(el, force) {
+  if (force === true) {
+    el.classList.remove('gt-hidden');
+    assertShown(el, true);
+  } else if (force === false) {
+    el.classList.add('gt-hidden');
+    assertShown(el, false);
+  } else if (force === undefined) {
+    const wasShown = window.config.runModeIsProd ? undefined : isShown(el);
+    el.classList.toggle('gt-hidden');
+    if (wasShown !== undefined) {
+      assertShown(el, !wasShown);
+    }
+  } else {
+    throw new Error('invalid force argument');
+  }
+}
```

Comment:
I'm absolutely confused by this logic, shouldn't it be
```suggestion
function toggleShown(el, visible /*optional*/) {
  if (visible === undefined) {
    const wasShown = window.config.runModeIsProd ? undefined : isShown(el); // Test only in development if the expected state has been reached
    el.classList.toggle('gt-hidden');
    if (wasShown !== undefined) {
      assertShown(el, !wasShown);
    }
  } else if (visible === true) {
    el.classList.remove('gt-hidden');
    assertShown(el, true);
  } else if (visible === false) {
    el.classList.add('gt-hidden');
    assertShown(el, false); 
  } else {
    throw new Error('invalid visible argument: ' + visible);
  }
}
```
?

Reply:
No. It's called by `force` from MDN

https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/toggle

Reply:
Interesting. What about 
```suggestion
function toggleShown(el, force) {
  if (force === undefined) {
    const wasShown = window.config.runModeIsProd ? undefined : isShown(el); // Test only in development if the expected state has been reached
    el.classList.toggle('gt-hidden');
    if (wasShown !== undefined) {
      assertShown(el, !wasShown);
    }
  } else if (force === true) {
    el.classList.remove('gt-hidden');
    assertShown(el, true);
  } else if (force === false) {
    el.classList.add('gt-hidden');
    assertShown(el, false);
  } else {
    throw new Error('invalid force argument: ' + force);
  }
}
```
then?

Reply:
But why? I like to put the defined values `true/false` at the beginning.

Reply:
Well… At least for me, it makes more sense to check first if the value is present, and then which of the accepted values it has.
Either way, that's what I certainly would like, independently from the order of values:
```suggestion
function toggleShown(el, force) {
  if (force === true) {
    el.classList.remove('gt-hidden');
    assertShown(el, true);
  } else if (force === false) {
    el.classList.add('gt-hidden');
    assertShown(el, false);
  } else if (force === undefined) {
    const wasShown = window.config.runModeIsProd ? undefined : isShown(el); // Test only in development if the expected state has been reached
    el.classList.toggle('gt-hidden');
    if (wasShown !== undefined) {
      assertShown(el, !wasShown);
    }
  } else {
    throw new Error('invalid force argument: ' + force);
  }
}
```

Reply:
But it makes more sense to me to check whether the user want to show or hide the element.

And I can find more reasons for this, in most cases, the caller is sure to show or hide, they do not need toggle behavior.

So, checking true/false at the beginning is more efficient, right?

Reply:
I do not see why the comment "// Test only in development if the expected state has been reached" is necessary.

I think the code is clear enough when reading about "window.config.runModeIsProd", it clear that the `isShown` is only called for non-prod, by design.


I do not to take the comment by myself. I am open and if other people like the comment, feel free to edit this PR directly.